### PR TITLE
REG-21 test edge case of cancelling donation and re-entering details

### PR DESCRIPTION
**Pull request by alihejazi**

### CODE REVIEW

**Comment by alihejazi** [comment]
This should be merged after https://github.com/thebiggive/regression-tests/pull/74 because it branches off of it.

**Comment by bdsl** [comment]
Sorry closed by mistake.

File: features/donation-as-new-donor-restart-donation.feature [hunk 1]
```diff
@@ -0,0 +1,23 @@
+Feature: Make a matched donation to a Stripe charity – new donor - edge case of restarting donation
```
Comment (bdsl) [inline features/donation-as-new-donor-restart-donation.feature:1]: Isn't this the same feature as "Make a matched donation to a Stripe charity – new donor" - just a different scenario?

I realise at the moment we only have one scenario for every feature, so not a blocker, but it might make sense to move this into the same feature file as the existing scenario - at least that seems to be how the tool is intended to be used, I don't know if there's a specific reason not to do that in our case.

File: pages/DonateStartPage.js [hunk 2]
```diff
@@ -299,6 +300,9 @@ export default class DonateStartPage {
      * Choose email communication preferences.
      */
     async setCommsPreferences() {
+        // Add some grace pause to allow enough time for the checkboxes to render
```
Comment (bdsl) [inline pages/DonateStartPage.js:303]: Can make this a bit more succinct - don't need to repeat the code in the comment.
```suggestion
        // Allow enough time for the checkboxes to render
```

File: steps/donation.js [hunk 3]
```diff
@@ -13,8 +13,13 @@ import CharityPortalLoginPage from '../pages/CharityPortalLoginPage';
...
 
-// Constants
-const donationAmount = randomIntFromInterval(5, 100);
+/**
+ * Note: donationAmount is changable in the `restart-donation` test, whereby the bot changes the
+ * donation amount after cancelling the first. Hence, it's not a constant variable.
+ * See `When('I re-enter an amount between £5 and £25,000', ...)` method.
+ * See REG-21
+ */
+let donationAmount = randomIntFromInterval(5, 100);
```
Comment (bdsl) [inline steps/donation.js:22]: I'm not sure, but it might be better to keep the const and have two of them in the restart test - I guess consts are a bit simpler to work with than vars

File: steps/donation.js [hunk 4]
```diff
@@ -98,6 +103,18 @@ When(
     }
 );
 
+When(
+    'I re-enter an amount between £5 and £25,000',
+    async () => {
+        // Update donation amount by -1, relative to its itinial value.
+        // Re-store new value in same variable so that the following check passes later:
+        // `DonateSuccessPage.checkBalance(donationAmount);`
+        donationAmount -= 1;
```
Comment (bdsl) [inline steps/donation.js:112]: How do we know that decrementing won't take out of the valid range of values?

Maybe we don't need to use random values and we can have a scenario with hard-coded values for the initial amount and the adjusted amount.

**Review by bdsl — APPROVED** [review]
Added some comments, no blockers but take a look and let me know what you think.

File: steps/donation.js (hunk 3 above)
Comment (alihejazi) [reply steps/donation.js:22]: @bdsl I think it's better to use one variable so we don't have to update the logic for the `'my last email should contain the correct amounts'` test:

```
Then(
    'my last email should contain the correct amounts',
    async () => {
        if (!(await checkAnEmailBodyContainsText(
            `Donation: <strong>£${donationAmount}.00</strong>`,
        ))) {
            throw new Error(`Donation amount £${donationAmount} not found in email`);
        }
    }
);
```

Comment (bdsl) [reply steps/donation.js:22]: Maybe - although if we kept `donationAmount` to mean the final donation amount and introduced a new variable for the first amount entered we wouldn't have to update that. Anyway not a blocker.

File: features/donation-as-new-donor-restart-donation.feature (hunk 1 above)
Comment (alihejazi) [reply features/donation-as-new-donor-restart-donation.feature]: @bdsl I think it's fine as a test on its own. Like you've identified, it's a different scenario, hence a different test file. The first tests a successful guest donation journey, and though this one seems similar enough, the restarting of a donation causes a cancellation of the donation in MatchBot and the generation of a new Payment Intent in Stripe.

**Comment by alihejazi** [comment]
Tests should all be passing in the next run.

**Review by bdsl — APPROVED** [review]
LGTM
